arroyo/processing/strategies/streaming/collect.py

```diff
@@ -97,54 +98,105 @@ def __init__(
+            logger.debug("Size limit reached, closing %r...", self.batch)
+            self.close_and_reset_batch()
+        elif self.batch.duration() >= self.__max_batch_time:
+            logger.debug("Time limit reached, closing %r...", self.batch)
+            self.close_and_reset_batch()
 
     def submit(self, message: Message[TPayload]) -> None:
         assert not self.__closed
 
-        if self.__batch is None:
-            self.__batch = Batch(self.__step_factory(), self.__commit_function)
+        if self.batch is None:
+            self.batch = Batch(self.__step_factory(), self.__commit_function)
 
-        self.__batch.submit(message)
+        self.batch.submit(message)
 
     def close(self) -> None:
         self.__closed = True
 
-        if self.__batch is not None:
-            logger.debug("Closing %r...", self.__batch)
-            self.__batch.close()
+        if self.batch is not None:
+            logger.debug("Closing %r...", self.batch)
+            self.batch.close()
 
     def terminate(self) -> None:
         self.__closed = True
 
-        if self.__batch is not None:
-            self.__batch.terminate()
+        if self.batch is not None:
+            self.batch.terminate()
 
     def join(self, timeout: Optional[float] = None) -> None:
-        if self.__batch is not None:
-            self.__batch.join(timeout)
-            logger.info("Completed processing %r.", self.__batch)
-            self.__batch = None
+        if self.batch is not None:
+            self.batch.join(timeout)
+            logger.info("Completed processing %r.", self.batch)
+            self.batch = None
+
+
+class ParallelCollectStep(CollectStep[TPayload]):
+    """
+    ParallelCollectStep is similar to CollectStep except it allows the closing and reset of the
+    batch to happen in a threadpool. What is allows is for the next batch to start getting
+    filled in while the previous batch is still being processed.
+
+    The threadpool will have only 1 worker since we want to perform writes to clickhouse sequentially
+    so that kafka offsets are written in order.
+    """
+
+    def __init__(
+        self,
+        step_factory: Callable[[], ProcessingStep[TPayload]],
+        commit_function: Callable[[Mapping[Partition, Position]], None],
+        max_batch_size: int,
+        max_batch_time: float,
+    ):
+        super().__init__(step_factory, commit_function, max_batch_size, max_batch_time)
+        self.__threadpool = ThreadPoolExecutor(max_workers=1)
+        self.future: Optional[Future[None]] = None
+
+    def close_and_reset_batch(self) -> None:
+        """
+        Closes the current batch in an asynchronous manner. Waits for previous work to be completed before proceeding
+        the next one. We can provide the existing batch to the threadpool since the collector is going to make a
+        new batch.
+        """
+        if self.future:
+            self.future.result()
```

Comment:
should this have a timeout at all? not sure how likely this is to raise a `TimeoutError`. Seems like we should def be done submitting the previous batch before a new batch is filled up, but maybe if the consumer is backlogged this could happen?

Reply:
Added a timeout of 5 seconds. Most often the future's result would be available immediately. But I have added it if for some reason clickhouse is down, we can abort.